src/lib/Microsoft.Health.Extensions.Fhir.R4/FhirClientValidator.cs

```diff
@@ -4,22 +4,28 @@
 // -------------------------------------------------------------------------------------------------
 
 using System;
+using System.Net.Http;
+using System.Threading.Tasks;
 using EnsureThat;
-using Hl7.Fhir.Rest;
 using Microsoft.Health.Extensions.Fhir.Telemetry.Exceptions;
+using Microsoft.Health.Fhir.Client;
 using Microsoft.Health.Logging.Telemetry;
 
 namespace Microsoft.Health.Extensions.Fhir
 {
-    public static class FhirServiceValidator
+    public static class FhirClientValidator
     {
-        public static bool ValidateFhirService(FhirClient client, ITelemetryLogger logger)
+        public static async Task<bool> ValidateFhirClientAsync(
+            this HttpClient client,
+            ITelemetryLogger logger)
         {
             EnsureArg.IsNotNull(client, nameof(client));
+            EnsureArg.IsNotNull(logger, nameof(logger));
 
             try
             {
-                client.CapabilityStatement(SummaryType.True);
+                var fhirClient = new FhirClient(client);
```

Comment:
We are building a new FhirClient here, and we don't really need to. You can adjust the original method signature and make this an extension method for `IFhirClient`. And invoke it as follows on the Singleton FhirClient that we have available via DI:

```c#
var serviceProvider = serviceCollection.BuildServiceProvider();
var fhirClient = serviceProvider.GetRequiredService<IFhirClient>();
var logger = serviceProvider.GetRequiredService<ITelemetryLogger>();

await fhirClient.ValidateFhirClientAsync(logger);
```

We just need a place to execute the above. Perhaps at the end of `ProcessorStartup#ConfigureServices` method?

Reply:
Thanks for reviewing Rob. This definitely works for the console project. However ProcessorStartup isn't called when running the functions (Microsoft.Health.Fhir.Ingest.Host project). If we're to place fhirClient.ValidateFhirClientAsync() at the top of both the Normalization and FhirTransformation functions, the validation method would run each time they're triggered.

We could consider calling await fhirClient.ValidateFhirClientAsync(logger) at the end of AddMeasurementFhirImport() and in ProcessorStartup.ConfigureServices. My thought is that it'd best to call the validation method when we configure the client though, so we don't have to call it in multiple locations.

Happy to discuss further

Reply:
Synced offline and pushed new changes